crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs

```diff
@@ -0,0 +1,184 @@
+use crate::checkers::ast::Checker;
+use crate::importer::ImportRequest;
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{CmpOp, Expr, ExprCompare, ExprStringLiteral};
+use ruff_text_size::{Ranged, TextRange};
+
+/// ## What it does
+/// Checks for uses of hardcoded charsets, which are defined in Python string module.
+///
+/// ## Why is this bad?
+/// Usage of named charsets from the standard library is more readable and less error-prone.
+///
+/// ## Example
+/// ```python
+/// x = "0123456789"
+/// y in "abcdefghijklmnopqrstuvwxyz"
+/// ```
+///
+/// Use instead
+/// ```python
+/// import string
+///
+/// x = string.digits
+/// y in string.ascii_lowercase
+/// ```
+///
+/// ## References
+/// - [Python documentation: String constants](https://docs.python.org/3/library/string.html#string-constants)
+#[violation]
+pub struct HardcodedStringCharset {
+    name: &'static str,
+}
+
+impl AlwaysFixableViolation for HardcodedStringCharset {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use of hardcoded string charset")
+    }
+
+    fn fix_title(&self) -> String {
+        let HardcodedStringCharset { name } = self;
+        format!("Replace hardcoded charset with `string.{name}`")
+    }
+}
+
+struct NamedCharset {
+    name: &'static str,
+    bytes: &'static [u8],
+    ascii_char_set: AsciiCharSet,
+}
+
+/// Represents the set of ascii characters in form of a bitset.
+#[derive(Copy, Clone, Eq, PartialEq)]
+struct AsciiCharSet(u128);
+
+impl AsciiCharSet {
+    /// Creates the set of ascii characters from `bytes`.
+    /// Returns None if there is non-ascii byte.
+    fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        bytes
+            .iter()
+            .try_fold(0, |acc, &byte| byte.is_ascii().then(|| acc | (1 << byte)))
+            .map(Self)
+    }
```

Comment:
What's the motivation for having both `from_bytes` and `from_bytes_const`? Aren't both returning `None` if the string contains any non ascii character?

Reply:
I would like to stick to using a `const` here (another option, I would like to avoid, is runtime evaluation under "OnceLock"):
https://github.com/astral-sh/ruff/blob/066e7006e4b5fad341fa8c84a85ec1f5cf5d3602/crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs#L108
So I need `from_bytes` to be a `const fn`, but unfortunately currently "in stable" it cannot have a clean implemenration like this:
https://github.com/astral-sh/ruff/blob/066e7006e4b5fad341fa8c84a85ec1f5cf5d3602/crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs#L62-L65

So I added the dirty implementation (which can be `const fn`) under another name `from_bytes_const`:
https://github.com/astral-sh/ruff/blob/066e7006e4b5fad341fa8c84a85ec1f5cf5d3602/crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs#L70-L85 

And there is a TODO to remove it, when stable supports `const fn` like usual `from_bytes`.

The third factory is about the usage of `Option::unwrap` in `const fn`. Currently is not possible (wait for https://github.com/rust-lang/rust/issues/67441). When it will be shipped, we can write
```rust
            ascii_char_set: AsciiCharSet::from_bytes_const(bytes).unwrap(),
```
instead of
https://github.com/astral-sh/ruff/blob/066e7006e4b5fad341fa8c84a85ec1f5cf5d3602/crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs#L103

Reply:
I do like the const version but do we need  the non const version?

Reply:
The const version is non-idiomatic and will surprise a reader.

Having this non-idiomatic version in runtime could have some interesting behavior of optimization part of compiler. 
While there is a difference in generated code ( see https://godbolt.org/z/53T3z86f8 ), I assume it is a good default decision to have two implemenation untill the `const fn` supports idiomatic implementation.

Reply:
I'd be fine with keeping the code simpler and using `OnceLock` if that works personally.

Otherwise, if the non-const version isn't being used, I'd drop it. If you're worried about surprising the reader, then you could write a short comment. (Although I don't think it's that surprising. If you've ever tried to write a const fn before, you'll know that you're pretty limited in the tools you have.)

Reply:
https://github.com/astral-sh/ruff/pull/13530/commits/b9b3b64937f000c66a850b541df17b4899bb3ed5
Ok. `AsciiCharSet` interface was simplified to the direction of only `const` implementation. 
Also: 
- `from_bytes_const_unwrap` was inlined to localize the story of `Option::unwrap` in `const` context.
- slightly change the `TODO` comments.